Sources/PackageCollectionGenerator/PackageMetadataProviders/GitLabPackageManagerMetadataProvider.swift

```diff
@@ -0,0 +1,170 @@
+            case (_, _, 0):
+                result = .failure(Errors.apiLimitsExceeded(metadataURL, apiLimit, apiRemaining))
+            case (401, true, _):
+                result = .failure(Errors.invalidAuthToken(metadataURL))
+            case (401, false, _):
+                result = .failure(Errors.permissionDenied(metadataURL))
+            case (403, _, _):
+                result = .failure(Errors.permissionDenied(metadataURL))
+            case (404, _, _):
+                result = .failure(Errors.notFound(metadataURL))
+            case (200, _, _):
+                guard let metadata = try? response.decodeBody(GetRepositoryResponse.self, using: self.decoder) else {
+                    callback(.failure(Errors.invalidResponse(metadataURL, "Invalid body")))
+                    return
+                }
+
+                let license = metadata.license
+                let packageLicense: PackageCollectionModel.V1.License?
+                if let license = license {
+                    packageLicense = .init(name: license.name, url: license.sourceURL)
+                } else if let licenseURL = metadata.license_url,
+                    let licenseURL = URL(string: licenseURL) {
+                    packageLicense = .init(name: nil, url: licenseURL)
+                } else {
+                    packageLicense = nil
+                }
+
+                let model = PackageBasicMetadata(
+                    summary: metadata.description,
+                    keywords: metadata.topics,
+                    readmeURL: metadata.readme_url.flatMap { URL(string: $0) },
+                    license: packageLicense
+                )
+
+                callback(.success(model))
+            default:
+                callback(.failure(Errors.invalidResponse(metadataURL, "Invalid status code: \(response.statusCode)")))
+            }
+        }
+    }
+
+    internal func apiURL(_ url: String) -> Foundation.URL? {
+        return URL(string: url.spm_dropGitSuffix())
+    }
+
+    private func makeRequestOptions(validResponseCodes: [Int]) -> HTTPClientRequest.Options {
+        var options = HTTPClientRequest.Options()
+        options.addUserAgent = true
+        options.validResponseCodes = validResponseCodes
+        options.authorizationProvider = { url in
+            url.host.flatMap { host in
+                return self.authTokens[.gitlab(host)].flatMap { token in "Bearer \(token)" }
+            }
+        }
+        return options
+    }
+
+    private static func makeDefaultHTTPClient() -> HTTPClient {
+        var client = HTTPClient()
+        client.configuration.requestTimeout = .seconds(2)
+        client.configuration.retryStrategy = .exponentialBackoff(maxAttempts: 3, baseDelay: .milliseconds(50))
+        client.configuration.circuitBreakerStrategy = .hostErrors(maxErrors: 50, age: .seconds(30))
+        return client
+    }
+
+    enum Errors: Error, Equatable {
```

Comment:
Perhaps refactor this into `PackageMetadataProviderError` so it can be shared across all providers.

Reply:
I only realise now that there is also code inside SwiftPM to fetch metadata. And it seems like that the usage of the `GitHubPackageMetadataProvider` is hard-coded in it. This seems like that there should be a bigger consolidation overall. I agree with you that the error-codes should be shared, but I would either put them inside a new file inside this repository or defer it for a greater consolodation inside SwiftPM. 
Is there anything planned to consolidate the metadata-provider-logic of SwiftPM and swift-package-collection-generator?

Reply:
> Is there anything planned to consolidate the metadata-provider-logic of SwiftPM and swift-package-collection-generator?

There is plan to move swift-package-collection-generator into SwiftPM actually. This repo started out separate from SwiftPM because we needed to be able to iterate on this more quickly, but now that things have stabilized, we would like to make this part of SwiftPM, which is more fitting. And when we do that, we will consolidate the code as you point out.

Let's just leave it as-is for now.